Add iterators to Queue, change coding style.

JDuchniewicz opened the pull request:
Adding iterator functionality to core::Queue, basing it on std::vector as underlying container

JDuchniewicz committed:
Add iterators to Queue, change coding style.

JDuchniewicz committed:
CI fixes

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -3,71 +3,66 @@
-
-		/// <summary>Creates queue instance with provided capacity.</summary>
+    template<class T>
+    class Queue : public BaseObject<>
```
Why change from LiteralType to normal? There is no need for Queue to have vtable.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -3,71 +3,66 @@
-		Queue() = default;
-
-		/// <summary>Creates queue instance with provided capacity.</summary>
+    template<class T>
```
We'd rather use typename instead of class for such cases.

JDuchniewicz replied:
with C++20 typename/class keywords will disapperar, but changing for the time being

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -3,71 +3,66 @@
+		Queue(const std::initializer_list<T>& list) { populateFromInitializerList(list); }
 
 		/// <summary>Basic destructor.</summary>
 		~Queue()
```
If the destructor is empty, remove it.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+	*/
 		/// <summary>Performs removal from the back of the queue.</summary>
-		void PopBack()
+		void popBack()
```
Let's make pop methods return the popped object. Just remember to move it out of the vector and not to copy.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
-			HEAVY_ASSERTE(capacity >= Capacity, "Capacity too small!");
-			Realloc(capacity);
+			HEAVY_ASSERTE(capacity >= m_capacity, "Capacity too small!");
+			realloc(capacity);
```
Why not `m_data.reserve(capacity);` ? Then you can remove `realloc(..)` completely.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
 		/// <param name="rhs">Searched object.</param>
 		/// <returns>True if present, false otherwise.</returns>
-		bool Contains(const T& rhs) const { return FindIdx(rhs) < GetSize(); }
+		bool contains(const T& rhs) const { return findIdx(rhs) < getSize(); }
```
Remove all find and contains methods. Let's force user to use `std::algorithm`

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+
+			Iterator& operator++() { idx = ref.getNextIdx(idx); return *this; }
+			Iterator operator++(int) { Iterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+			Iterator& operator--() { idx = ref.getNextIdx(idx); return *this; }
```
Probably should be `getPrevIdx` same for the other operator. Let's make sure we have as much test coverage as possible for this class. We should aim for around 100%.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+			bool operator!=(const Iterator& rhs) const { return idx != rhs.idx; }
+			Iterator& operator=(const Iterator& rhs) = default;
+
+			T& operator*() const { return ref.m_data.at(idx); }
```
What about `operator->` ? Do we get it for free?

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
-		bool Contains(const T& rhs) const { return FindIdx(rhs) < GetSize(); }
+		bool contains(const T& rhs) const { return findIdx(rhs) < getSize(); }
+		
+		class Iterator : public BaseObject<>
```
all iterators should be BaseObjectLiteralType, the inheritance should be final.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+
+			ConstIterator& operator++() { idx = ref.getNextIdx(idx); return *this; }
+			ConstIterator operator++(int) { ConstIterator ret(ref, idx); idx = ref.getNextIdx(idx); return ret; }
+			ConstIterator& operator--() { idx = ref.getNextIdx(idx); return *this; }
```
Change to `getPrevIdx` for both operators --

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+
+			ReverseIterator& operator++() { increment(); return *this; }
+			ReverseIterator operator++(int) { ReverseIterator ret(ref, idx); increment(); return ret; }
+			ReverseIterator& operator--() { idx = ref.getPrevIdx(idx); return *this; }
```
Since this is `ReverseIterator` `operator--` should call `getNextIdx`. Same for `ConstReverseIterator`

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
-		void DecreaseIdx(size_t& idx) const { idx = GetPrevIdx(idx); }
-		void AdvanceIdx(size_t& idx) const { idx = GetNextIdx(idx); }
-		size_t GetNthIdx(size_t& n)	const { return (Head + n) % Capacity; }
+		size_t getPrevIdx(size_t idx) const { return (idx == 0) ? m_capacity - 1 : idx - 1; }
```
Just thought of optimization, change this line to `return (m_capacity + idx - 1) % m_capacity;`

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+					idx = ref.m_tail;
+				else 
+					idx = ref.getPrevIdx(idx);
+			}
```
The way to implement interators in a circular buffer is too always keep two elements in the queue:
- element before head - which is == queue.rend()
- element after end - which is == queue.end()
I'm not sure whether the iterators work correctly in the current setup. For sure we need good test coverage here.

JDuchniewicz replied:
Currently there is a workaround implemented, please review if it is suitable. It is hidden from outside world, and test do cover it.

JDuchniewicz replied:
proper solution implemented in newest commits

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
 
 		//------------------------------------------------------------------------------
-		size_t FindIdx(const T& rhs) const
+		size_t findIdx(const T& rhs) const
```
Remove this method.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
 
 		//------------------------------------------------------------------------------
-		void Realloc(size_t capacity)
+		void realloc(size_t capacity)
```
This method is now broken. What it did before and now doesn't is defragment the queue. Which basically means to copy elements in such a way to make m_head == 0 and m_end == size.

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+		std::vector<T> m_data;
+    };
+	
+	template<typename T> typename ::pe::core::storage::Queue<T>::Iterator begin(::pe::core::storage::Queue<T>& rhs) { return rhs.begin(); }
```
Those functions are no longer needed. Remove them.

MuniuDev reviewed (changes requested):
Nice overhaul. There are some problems though:
- please fix the indentation. It seems that you're using spaces instead of tabs.
- Please add a lot more unit tests with a lot more tests cases (divide large TEST_CASES, to a lot more smaller ones with one scenario each). You can also try using test sections functionallity in catch.

JDuchniewicz committed:
Revert to original managed buffer

JDuchniewicz committed:
Add more test cases. Fix iteration bug proposal

This removes need for additional element added to queue in order to enable reverse iteration capabilities

JDuchniewicz committed:
Test split into smaller units

JDuchniewicz committed:
Change underlying container

JDuchniewicz committed:
Add additional increment/decrement operators for queue cells. Add breaking test for wrapping

MuniuDev committed:
Merge branch 'dev' into stlqueue

MuniuDev committed:
Queue fixes

JDuchniewicz committed:
Add more tests, restore old copy-truncate behaviour of queue

JDuchniewicz committed:
Merge branch 'dev' of https://github.com/KNTGPolygon/PolyEngine into stlqueue

JDuchniewicz committed:
Merge branch 'dev' of https://github.com/KNTGPolygon/PolyEngine into stlqueue

JDuchniewicz committed:
CI fixes for MSVC

MuniuDev commented on PolyEngine/Core/Src/pe/core/storage/Queue.hpp:
```diff
@@ -78,198 +77,337 @@ namespace pe::core::storage
+		{
+			// PhysicalSize takes guard cell into account
+			const size_t physicalSize = m_size + 1;
+			idx = (idx + 1 - m_head) % physicalSize;
```
we can still get an overflow here if the head is bigger than idx + 1. Add physicalSize to idx as is done in getPrevIdx.

JDuchniewicz committed:
Fix wrapped queue iteration

JDuchniewicz committed:
Merge branch 'dev' into stlqueue

JDuchniewicz committed:
Merge branch 'dev' into stlqueue

JDuchniewicz committed:
Use new buffer iteration functions
